### PR TITLE
Add conditions to DataVolume and add condition annotation to PVCs

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
Add conditions to data volumes so we can derive the status of the datavolume from them.
There are 3 conditions:
1. ReadyCondition, if true, the DataVolume is ready.
2. BoundCondition, if true, the underlying PVC has been bound.
3. RunningCondition, if true, the population step of the DataVolume is still running. This usually means a pod is working on import/upload/clone.

All 3 conditions can go from true to false and vice versa based on what is happening in the cluster. The running and ready conditions are mutually exclusive. You cannot be ready and running at the same time.

Each condition has 6 fields:
- Type
- Status
- Reason
- Message
- LastTransitionTime
- LastHeartBeatTime

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Enhancement: DataVolumes now have 3 conditions, Ready, Bound, and Running.
```

### CODE REVIEW

**Comment by awels** [comment]
/hold
I need to add documentation and reason codes.

File: pkg/apis/core/v1alpha1/types.go [hunk 1]
```diff
@@ -141,12 +143,25 @@ type DataVolumeList struct {
 	Items []DataVolume `json:"items"`
 }
 
+// DataVolumeCondition represents the state of a data volume condition.
+type DataVolumeCondition struct {
```
Comment (mhenriks) [inline pkg/apis/core/v1alpha1/types.go:147]: The timestamp fields on most conditions are missing `lastTransitionTime` and `lastProbeTime`/`LastHeartbeatTime`

File: pkg/apis/core/v1alpha1/types.go [hunk 2]
```diff
@@ -141,12 +143,25 @@ type DataVolumeList struct {
...
+	Type   DataVolumeConditionType `json:"type" description:"type of condition ie. Ready|Bound|Running."`
+	Status corev1.ConditionStatus  `json:"status" description:"status of the condition, one of True, False, Unknown"`
+	// +optional
+	Reason string `json:"reason,omitempty" description:"one-word CamelCase reason for the condition's last transition"`
```
Comment (mhenriks) [inline pkg/apis/core/v1alpha1/types.go:151]: Why only one word?

File: pkg/apis/core/v1alpha1/types.go (hunk 1 above)
Comment (awels) [reply pkg/apis/core/v1alpha1/types.go:147]: I purposely didn't add transition/lastprobe/heartbeart/other time. After doing some research I noticed basically the only conditions that used that was the conditions we put in for HCO. The time didn't add much/any value so why bother adding it?

File: pkg/apis/core/v1alpha1/types.go (hunk 2 above)
Comment (awels) [reply pkg/apis/core/v1alpha1/types.go:151]: So I can easily copy to/from event reasons without having to modify anything.

File: pkg/apis/core/v1alpha1/types.go (hunk 1 above)
Comment (mhenriks) [reply pkg/apis/core/v1alpha1/types.go:147]: I think there is value.  At least in transition time.  And I've never encountered conditions without timestamps.  Here are some places that do:

https://github.com/kubevirt/kubevirt/blob/master/staging/src/kubevirt.io/client-go/api/v1/types.go#L266-L284

https://github.com/kubevirt/kubevirt/blob/master/staging/src/kubevirt.io/client-go/api/v1/types.go#L664-L671

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L448-L460

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2250-L2262

https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L4009-L4021

File: pkg/apis/core/v1alpha1/types.go (hunk 2 above)
Comment (mhenriks) [reply pkg/apis/core/v1alpha1/types.go:151]: That sounds easy but maybe more detailed info would be helpful?

Comment (awels) [reply pkg/apis/core/v1alpha1/types.go:151]: Well the message has more detail.

File: pkg/apis/core/v1alpha1/types.go (hunk 1 above)
Comment (awels) [reply pkg/apis/core/v1alpha1/types.go:147]: Okay, I have added transition time.

File: pkg/apis/core/v1alpha1/types.go [hunk 3]
```diff
@@ -129,6 +129,8 @@ type DataVolumeStatus struct {
 	Phase        DataVolumePhase    `json:"phase,omitempty"`
 	Progress     DataVolumeProgress `json:"progress,omitempty"`
 	RestartCount int32              `json:"restartCount"`
+	// +listType=set
+	Conditions []*DataVolumeCondition `json:"conditions,omitempty" optional:"true"`
```
Comment (mhenriks) [inline pkg/apis/core/v1alpha1/types.go:133]: Unusual to have slice of pointers in k8s api.  `Conditions []DataVolumeCondition` preferred

Comment (awels) [reply pkg/apis/core/v1alpha1/types.go]: Yeah, was having issue with getting to the backing data of the slice, for some reason this:
```go
for _, cond := range conditions {
  return &cond
}
```
doesn't give me the backing data, but
```go
for i, cond := range conditions {
  return &conditions[i]
}
```
does

Comment (mhenriks) [reply pkg/apis/core/v1alpha1/types.go]: In your first example, `cond` is a copy of the slice item

Comment (awels) [reply pkg/apis/core/v1alpha1/types.go]: Yeah I had initially not realized this, that is why I did the pointer, that way the pointer was the copy and I was able to modify the actual struct. Once I realized its a copy, and I could return the pointer in the second way there was no reason for the pointer slice.

File: pkg/controller/datavolume-conditions.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+			Type: conditionType,
+		})
+		condition = findConditionByType(conditionType, conditions)
+	}
+	if condition.Status != status {
+		condition.LastTransitionTime = metav1.Now()
+	}
+	condition.Status = status
+	condition.Message = message
+	condition.Reason = reason
```
Comment (mhenriks) [inline pkg/controller/datavolume-conditions.go:58]: I think we should also only update message/reason when the status changes

File: pkg/controller/datavolume-conditions.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		conditions = append(conditions, cdiv1.DataVolumeCondition{
+			Type: cdiv1.DataVolumeRunning,
+		})
+		condition = findConditionByType(cdiv1.DataVolumeRunning, conditions)
+	}
+	if val, ok := anno[AnnRunningConditionMessage]; ok {
+		condition.Message = val
+	} else {
+		condition.Message = ""
+	}
```
Comment (mhenriks) [inline pkg/controller/datavolume-conditions.go:74]: above if/else is equivalent to `val := anno[AnnRunningConditionMessage]`

Comment (awels) [reply pkg/controller/datavolume-conditions.go:74]: Hmm I thought it would error if the key didn't exist, but it just return blank. will fix.

File: pkg/controller/datavolume-conditions.go (hunk 4 above)
Comment (awels) [reply pkg/controller/datavolume-conditions.go:58]: I went back and forth on this, and decided on always updating because its possible to stay at the same condition but have a different message/reason, I thought it was important to always have the latest message/reason, but if I do that, I should have the last transition time also be dependent on reason/message being different.

File: pkg/controller/datavolume-conditions.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+)
+
+const (
+	boundFalse      = "BoundChangeFalse"
+	transferRunning = "TransferRunning"
+	pvcBound        = "PVCBound"
+	pvcPending      = "PVCPending"
+	claimLost       = "ClaimLost"
+	notFound        = "NotFound"
```
Comment (mhenriks) [inline pkg/controller/datavolume-conditions.go:33]: Not clear to me why for example `PVCPending` is preferable to `PVC is pending` for condition reason

File: pkg/controller/datavolume-conditions.go (hunk 4 above)
Comment (mhenriks) [reply pkg/controller/datavolume-conditions.go:58]: I think the original reason is more important and don't want to lose it

File: pkg/controller/import-controller.go [hunk 7]
```diff
@@ -58,6 +58,12 @@ const (
 	AnnRequiresScratch = AnnAPIGroup + "/storage.import.requiresScratch"
 	// AnnDiskID provides a const for our PVC diskId annotation
 	AnnDiskID = AnnAPIGroup + "/storage.import.diskId"
+	// AnnRunningCondition provides a const for the running condition
+	AnnRunningCondition = AnnAPIGroup + "/storage.condition.running"
+	// AnnRunningConditionMessage provides a const for the running condition
+	AnnRunningConditionMessage = AnnAPIGroup + "/storage.condition.running.message"
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:64]: maybe `AnnLastTerminationMessage = AnnAPIGroup + "/storage.condition.termination.message"` makes more sense?

File: pkg/controller/import-controller.go [hunk 8]
```diff
@@ -243,12 +254,12 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
...
 			r.recorder.Event(pvc, corev1.EventTypeWarning, ErrImportFailedPVC, pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message)
+			anno[AnnRunningCondition] = "false"
+			anno[AnnRunningConditionMessage] = pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message
+			anno[AnnRunningConditionReason] = pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Reason
 		}
 	}
 
-	if pod.Status.ContainerStatuses != nil {
-		anno[AnnPodRestarts] = strconv.Itoa(int(pod.Status.ContainerStatuses[0].RestartCount))
-	}
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:251]: Why was this deleted?

File: pkg/controller/smart-clone-controller.go [hunk 9]
```diff
@@ -239,11 +239,17 @@ func (r *SmartCloneReconciler) updateSmartCloneStatusPhase(phase cdiv1.DataVolum
...
+		dataVolume.Status.Conditions = updateBoundCondition(dataVolume.Status.Conditions, newPVC)
+		dataVolume.Status.Conditions = updateReadyCondition(dataVolume.Status.Conditions, corev1.ConditionFalse, "", "")
+		dataVolume.Status.Conditions = updateCondition(dataVolume.Status.Conditions, cdiv1.DataVolumeRunning, corev1.ConditionTrue, MessageSmartClonePVCInProgress, SmartClonePVCInProgress)
 	case cdiv1.Succeeded:
 		dataVolumeCopy.Status.Phase = cdiv1.Succeeded
 		event.eventType = corev1.EventTypeNormal
 		event.reason = CloneSucceeded
 		event.message = fmt.Sprintf(MessageCloneSucceeded, dataVolumeCopy.Spec.Source.PVC.Namespace, dataVolumeCopy.Spec.Source.PVC.Name, newPVC.Namespace, newPVC.Name)
+		dataVolume.Status.Conditions = updateBoundCondition(dataVolume.Status.Conditions, newPVC)
+		dataVolume.Status.Conditions = updateReadyCondition(dataVolume.Status.Conditions, corev1.ConditionTrue, "", "")
```
Comment (mhenriks) [inline pkg/controller/smart-clone-controller.go:251]: Why no reason/message for ready?

File: pkg/controller/util.go [hunk 10]
```diff
@@ -296,3 +300,23 @@ func podPhaseFromPVC(pvc *v1.PersistentVolumeClaim) v1.PodPhase {
 func podSucceededFromPVC(pvc *v1.PersistentVolumeClaim) bool {
 	return (podPhaseFromPVC(pvc) == v1.PodSucceeded)
 }
+
+func setConditionFromPod(anno map[string]string, pod *v1.Pod) {
```
Comment (mhenriks) [inline pkg/controller/util.go:304]: This cannot be used for upload?

File: pkg/controller/import-controller.go (hunk 7 above)
Comment (awels) [reply pkg/controller/import-controller.go:64]: Sounds good

File: pkg/controller/import-controller.go (hunk 8 above)
Comment (awels) [reply pkg/controller/import-controller.go:251]: its part of setConditionFromPod now

File: pkg/controller/smart-clone-controller.go (hunk 9 above)
Comment (awels) [reply pkg/controller/smart-clone-controller.go:251]: Because the Ready Condition is mutually exclusive with the Running Condition, and I wanted the messages to come from the running condition.

File: pkg/controller/util.go (hunk 10 above)
Comment (awels) [reply pkg/controller/util.go:304]: Its because I moved
```
anno[AnnPodRestarts] = strconv.Itoa(int(pod.Status.ContainerStatuses[0].RestartCount))
```
In here, which is different for upload. But I think that is a mistake, if I leave it out I can re-use the rest in upload

File: pkg/controller/import-controller.go (hunk 8 above)
Comment (awels) [reply pkg/controller/import-controller.go:251]: Which is probably not a good plan, so reverting that.

File: pkg/controller/import-controller.go [hunk 11]
```diff
@@ -59,6 +59,12 @@ const (
 	AnnRequiresScratch = AnnAPIGroup + "/storage.import.requiresScratch"
 	// AnnDiskID provides a const for our PVC diskId annotation
 	AnnDiskID = AnnAPIGroup + "/storage.import.diskId"
+	// AnnRunningCondition provides a const for the running condition
+	AnnRunningCondition = AnnAPIGroup + "/storage.condition.running"
+	// AnnLastTerminationMessage provides a const for the running condition
+	AnnLastTerminationMessage = AnnAPIGroup + "/storage.condition.running.message"
+	// AnnRunningConditionReason provides a const for the running condition
+	AnnRunningConditionReason = AnnAPIGroup + "/storage.condition.termination.message"
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:67]:
```	
AnnLastTerminationMessage = AnnAPIGroup + "/storage.condition.termination.message"

AnnLastTerminationReason = AnnAPIGroup + "/storage.condition.termination.reason"
```

Comment (awels) [reply pkg/controller/import-controller.go]: Done

File: pkg/controller/import-controller.go [hunk 12]
```diff
@@ -244,12 +255,16 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
 			anno[AnnRequiresScratch] = "true"
 		} else {
 			r.recorder.Event(pvc, corev1.EventTypeWarning, ErrImportFailedPVC, pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message)
+			anno[AnnRunningCondition] = "false"
+			anno[AnnLastTerminationMessage] = pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message
+			anno[AnnLastTerminationReason] = pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Reason
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:260]: is this redundant with `setConditionFromPod` called above?

File: pkg/controller/import-controller.go [hunk 13]
```diff
@@ -261,6 +276,9 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
 				return err
 			}
 		}
+		anno[AnnRunningCondition] = "false"
+		anno[AnnLastTerminationMessage] = "Creating scratch space"
+		anno[AnnLastTerminationReason] = creatingScratch
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:281]: These are set when the pod is pending and needs scratch space.  But when the pod is no longer pending, it will display the real message/reason.  Is that okay?

Also, do any import types start out with scratch space?  If so it will never make sense to do this.

File: pkg/controller/upload-controller.go [hunk 14]
```diff
@@ -289,12 +291,18 @@ func (r *UploadReconciler) getOrCreateUploadPod(pvc *v1.PersistentVolumeClaim, p
...
 	scratchPvc := &corev1.PersistentVolumeClaim{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: pvc.Namespace}, scratchPvc); err != nil {
 		if !k8serrors.IsNotFound(err) {
 			return nil, errors.Wrap(err, "error getting scratch PVC")
 		}
 
+		anno[AnnRunningCondition] = "false"
+		anno[AnnLastTerminationMessage] = "Creating scratch space"
+		anno[AnnLastTerminationReason] = creatingScratch
+
```
Comment (mhenriks) [inline pkg/controller/upload-controller.go:305]: Not sure this ever makes sense.  Uploads should always have a scratch PVC and clones should never, right?

File: pkg/controller/import-controller.go (hunk 13 above)
Comment (awels) [reply pkg/controller/import-controller.go:281]: There are certain import types we know will have scratch space, in particular registry import.

File: pkg/controller/util.go [hunk 15]
```diff
@@ -296,3 +320,47 @@ func podPhaseFromPVC(pvc *v1.PersistentVolumeClaim) v1.PodPhase {
 func podSucceededFromPVC(pvc *v1.PersistentVolumeClaim) bool {
 	return (podPhaseFromPVC(pvc) == v1.PodSucceeded)
 }
+
+func setConditionFromPodWithPrefix(anno map[string]string, prefix string, pod *v1.Pod) {
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/util.go:324]: Done

File: pkg/controller/clone-controller.go [hunk 16]
```diff
@@ -297,7 +300,13 @@ func (r *CloneReconciler) validateSourceAndTarget(targetPvc *corev1.PersistentVo
 		return err
 	}
 
-	return ValidateCanCloneSourceAndTargetSpec(&sourcePvc.Spec, &targetPvc.Spec)
+	err = ValidateCanCloneSourceAndTargetSpec(&sourcePvc.Spec, &targetPvc.Spec)
+	if err == nil {
+		// Validation complete, put source PVC bound status in annotation
+		setBoundConditionFromPVC(targetPvc.GetAnnotations(), AnnSourceRunningCondition, sourcePvc)
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/clone-controller.go:306]: That is a mistake should be AnnBoundCondition

File: pkg/controller/datavolume-controller_test.go [hunk 17]
```diff
@@ -351,6 +373,39 @@ var _ = Describe("Reconcile Datavolume status", func() {
...
+func boundMessageByPVCPhase(pvcPhase corev1.PersistentVolumeClaimPhase) string {
+	if pvcPhase == corev1.ClaimBound {
+		return "PVC Bound"
+	} else if pvcPhase == corev1.ClaimPending {
+		return "PVC Pending"
+	} else if pvcPhase == corev1.ClaimLost {
+		return "Claim Lost"
+	}
+	return "No PVC found"
+}
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/datavolume-controller_test.go:396]: Done

File: pkg/controller/datavolume-controller.go [hunk 18]
```diff
@@ -631,9 +631,38 @@ func (r *DatavolumeReconciler) reconcileDataVolumeStatus(dataVolume *cdiv1.DataV
...
+	}
+
+	if pvc != nil {
+		anno = pvc.Annotations
+	} else {
+		anno = make(map[string]string, 0)
+	}
+
+	readyStatus := corev1.ConditionUnknown
+	switch phase := dataVolume.Status.Phase; phase {
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/datavolume-controller.go:652]: No clue, probably copy and paste from somewhere else.

File: pkg/controller/datavolume-conditions.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+		conditions = updateCondition(conditions, cdiv1.DataVolumeRunning, corev1.ConditionUnknown, anno[AnnRunningConditionMessage], anno[AnnRunningConditionReason])
+	}
+	return conditions
+}
+
+func updateReadyCondition(conditions []cdiv1.DataVolumeCondition, status corev1.ConditionStatus, message, reason string) []cdiv1.DataVolumeCondition {
+	return updateCondition(conditions, cdiv1.DataVolumeReady, status, message, reason)
+}
+
+func updateBoundCondition(conditions []cdiv1.DataVolumeCondition, pvc *corev1.PersistentVolumeClaim) []cdiv1.DataVolumeCondition {
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/datavolume-conditions.go:111]: Well the annotation can also be the bound status of the source PVC, if we are cloning. But it might be a good idea to put the logic in the respective controller instead of the data volume controller, and have the data volume controller just read the annotation.

File: pkg/controller/datavolume-conditions.go [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+		condition.Message = message
+		condition.Reason = reason
+		condition.LastHeartbeatTime = metav1.Now()
+	}
+	condition.Status = status
+	return conditions
+}
+
+func updateRunningCondition(conditions []cdiv1.DataVolumeCondition, anno map[string]string) []cdiv1.DataVolumeCondition {
+	if val, ok := anno[AnnRunningCondition]; ok {
```
[the remark this answers is not in the document]
Comment (awels) [reply pkg/controller/datavolume-conditions.go:69]: Well I can replace some if/else with a switch to make it more readable, but I have to calculate combinations, I guess I can split it out into several functions to make it clearer?

File: pkg/controller/import-controller.go [hunk 21]
```diff
@@ -261,7 +267,16 @@ func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, p
 				return err
 			}
 		}
+		anno[AnnBoundCondition] = "false"
```
Comment (mhenriks) [inline pkg/controller/import-controller.go:270]: Isn't it kind of weird that we don't actually check that the PVC is bound?  Rather assume based on the fact that the pod is pending and scratch space is required?  Could be pending for other reasons.

File: pkg/controller/upload-controller.go [hunk 22]
```diff
@@ -297,11 +301,18 @@ func (r *UploadReconciler) getOrCreateScratchPvc(pvc *v1.PersistentVolumeClaim,
...
+		anno[AnnBoundConditionReason] = creatingScratch
 		// Scratch PVC doesn't exist yet, create it.
 		scratchPvc, err = CreateScratchPersistentVolumeClaim(r.client, pvc, pod, name, storageClassName)
 		if err != nil {
 			return nil, err
 		}
+	} else {
+		delete(anno, AnnBoundCondition)
+		delete(anno, AnnBoundConditionMessage)
+		delete(anno, AnnBoundConditionReason)
```
Comment (mhenriks) [inline pkg/controller/upload-controller.go:315]: Not actually checking that the PVC is bound or not only it's existence

File: pkg/controller/import-controller.go (hunk 21 above)
Comment (awels) [reply pkg/controller/import-controller.go]: Good point, updated to check scratch PVC status, and put this condition in a different function.

File: pkg/controller/upload-controller.go (hunk 22 above)
Comment (awels) [reply pkg/controller/upload-controller.go]: Good point fixed.
